### DISCUSSION TITLE
Issues with Cefsharp dll missing in unit test project

### DISCUSSION THREAD

**Discussion by assopri** [cefsharp/CefSharp#3764] [General]
Hello! I use CefSharp Winforms 91 version. When I create unittest, using for the project using cefsharp, I get the following exception:

![image](https://user-images.githubusercontent.com/73235464/130029402-501d205b-7869-4cdc-8b10-e4707db79288.png)

It occurs in the moment when I add previously create browser object to the tabpage. At the same time, if I just create any other project instead of unit test - I get no issues with it.

By the way, I checked - dll is present in the corresponding folder:

![image](https://user-images.githubusercontent.com/73235464/130030262-409a3c2b-7b7c-44f1-a972-0fa85c1101c2.png)

**Comment by amaitland** [comment]
There are multiple unit test frameworks, it's not clear which you are using. 

CefSharp uses xunit for testing which you can use as a reference. The key parts are

- https://github.com/cefsharp/CefSharp/blob/master/CefSharp.Test/BindingRedirectAssemblyResolver.cs#L16
- https://github.com/cefsharp/CefSharp/blob/master/CefSharp.Test/CefSharpXunitTestFramework.cs
- https://github.com/cefsharp/CefSharp/blob/master/CefSharp.Test/CefSharpFixture.cs
- https://github.com/cefsharp/CefSharp/blob/master/CefSharp.Test/app.config#L7

**Reply by assopri** [reply]
Hi! I was using standard VS 2019 unit test project:
![image](https://user-images.githubusercontent.com/73235464/130196841-f307b8ea-cf91-4aa9-9837-381c5ba9c9d1.png)

**Reply by amaitland** [reply]
You need to disable running tests in app domains. It looks like mstest supports disabling app domains, at least based on https://github.com/microsoft/testfx/commit/c5184f9e11eba287f89813a761c9cc14cf4a09fc#diff-2da5904da633ce55977df086d0b0ccf6fa909a67235a324f535e5d04da223a1cR273

You can try getting mstest to copy the dependencies via https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.testtools.unittesting.deploymentitemattribute?view=visualstudiosdk-2019 or use a similar assembly resolve as linked above.

I haven't tried CefSharp with mstest so I don't know the exact steps required. Personally I prefer xunit.

**Reply by assopri** [reply]
Thank you for reply! Maybe you can give some tips how can I apply this approach microsoft/testfx@c5184f9#diff-2da5904da633ce55977df086d0b0ccf6fa909a67235a324f535e5d04da223a1cR273 in my case ? Just adding 2 methods (mentioned in this link), I get just more errors:
![image](https://user-images.githubusercontent.com/73235464/130757018-fc2fbdd5-c5f0-4bca-9e09-06f85d39a755.png)

I understand the question is pretty primitive, so I mean - what exactly parts of code I should take from that example? Sorry in advance, for this question

**Reply by assopri** [reply]
By the way, maybe my problem is because MSTest.exe is 32 bit and I have x64 app

**Reply by amaitland** [reply]
You'll need to add a runsettings file and set DisableAppDomain true. You don't actually add any of the code, it was purely for reference. 

https://docs.microsoft.com/en-us/visualstudio/test/configure-unit-tests-by-using-a-dot-runsettings-file?view=vs-2019

> By the way, maybe my problem is because MSTest.exe is 32 bit and I have x64 app

https://docs.microsoft.com/en-us/visualstudio/test/run-a-unit-test-as-a-64-bit-process?view=vs-2019 suggests you can run 64bit unit tests.

As I said before I've not tried mstest so I don't have any specific examples.
